### PR TITLE
Linear equality constraints

### LINKED ISSUE #25: Add support for linear equality constraints
@salomonl

### PR DESCRIPTION
**Pull request by salomonl**
Fix #25 
This pull request adds linear equality constraints for NOMAD.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -4,8 +4,11 @@ authors = ["Alexis Montoison (@amontoison), Pierrick Pascal (@ppascal97), Ludovi
...
 
 [deps]
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 NOMAD_jll = "2fc7fd02-576c-576e-a36a-6afce71715da"
+SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+Tulip = "6dd1b50a-3aae-11e9-10b5-ef983d2400fa"
 
 [compat]
 julia = "^1.3"
```
Comment (amontoison) [inline Project.toml:14]: We should add a compat entry for Tulip.
`Tulip = "^0.7.1"` ?

File: docs/src/tutorial.md [hunk 2]
```diff
@@ -69,3 +69,67 @@ p.options.max_bb_eval = 1000
...
 ```
+
+## Linear equality constraints
+
+It is equally possible to define linear equality constraints on a NomadProblem. In this case,
+NOMAD will solve the original problem on a reduced subspace of the original problem, via
+a conversion. For more details, the reader is invited to refer to:
+
+C. Audet, S. Le Digabel and M. Peyrega, Linear equalities in blackbox optimization.
+*Computational Optimization and Applications*, 61(1), 1-23, May 2015.
```
Comment (amontoison) [inline docs/src/tutorial.md:80]: Add an hyperlink with the DOI https://doi.org/10.1007/s10589-014-9708-2

File: docs/src/tutorial.md [hunk 3]
```diff
@@ -69,3 +69,67 @@ p.options.max_bb_eval = 1000
 result = solve(p, [0.0;2.0])
 println("Solution: ", result.x_best_feas)
 ```
+
+## Linear equality constraints
+
+It is equally possible to define linear equality constraints on a NomadProblem. In this case,
+NOMAD will solve the original problem on a reduced subspace of the original problem, via
+a conversion. For more details, the reader is invited to refer to:
```
Comment (amontoison) [inline docs/src/tutorial.md:77]:
```suggestion
It is also possible to define linear equality constraints in a NomadProblem structure.
NOMAD handles differently these constraints and solves the original problem on a reduced subspace.
For more details we refer the reader to:
```

File: docs/src/tutorial.md [hunk 4]
```diff
@@ -69,3 +69,67 @@ p.options.max_bb_eval = 1000
...
+      -0.2300117084673765077695861691609]
+
+# Solution
+result = solve(p, x0)
+println("Solution: ", result.x_best_feas)
+println("Satisfy Ax = b: ", A * result.x_best_feas ≈ b)
+println("And inside bound constraints: ", all(-10.0 .<= result.x_best_feas .<= 10.0))
+```
+
+The reader can take a look at the `test` folder for more difficult examples.
```
Comment (amontoison) [inline docs/src/tutorial.md:135]: J'ai l'exemple que ne fonctionne pas lorsque je le teste en local et il y a un problème avec la documentation de `nomadParameters` et `nomadParameters` : 
```julia
julia> include("make.jl")
[ Info: SetupBuildDirectory: setting up build directory.
[ Info: Doctest: running doctests.
[ Info: ExpandTemplates: expanding markdown templates.
┌ Warning: undefined binding 'nomadParameters' in `@docs` block in src/nomadParameters.md:5-7
│ ```@docs
│ nomadParameters
│ ```
└ @ Documenter.Expanders ~/.julia/packages/Documenter/FuXcO/src/Expanders.jl:298
┌ Warning: undefined binding 'nomadResults' in `@docs` block in src/nomadResults.md:5-7
│ ```@docs
│ nomadResults
│ ```
└ @ Documenter.Expanders ~/.julia/packages/Documenter/FuXcO/src/Expanders.jl:298
┌ Warning: failed to run `@example` block in src/tutorial.md:95-133
│ ```@example HS48
│ 
│ # blackbox
│ function bb(x)
│   f = (x[1]- 1)^2 * (x[2] - x[3])^2 + (x[4] - x[5])^2
│   bb_outputs = [f]
│   success = true
│   count_eval = true
│   return (success, count_eval, bb_outputs)
│ end
│ 
│ # linear equality constraints
│ A = [1.0 1.0 1.0 1.0 1.0;
│      0.0 0.0 1.0 -2.0 -2.0]
│ b = [5.0; -3.0]
│ 
│ # Define blackbox
│ p = NomadProblem(5, 1, ["OBJ"], # the equality constraints are not counted in the outputs of the blackbox
│                  bb,
│                  lower_bound = -10.0 * ones(5),
│                  upper_bound = 10.0 * ones(5),
│                  A = A, b = b)
│ 
│ # Fix some options
│ p.options.max_bb_eval = 500
│ 
│ # Define starting points. It must satisfy A x = b.
│ x0 = [0.57186958424864897665429452899843;
│       4.9971472653643420613889247761108;
│       -1.3793445664086618762667058035731;
│       1.0403394252630473459930726676248;
│       -0.2300117084673765077695861691609]
│ 
│ # Solution
│ result = solve(p, x0)
│ println("Solution: ", result.x_best_feas)
│ println("Satisfy Ax = b: ", A * result.x_best_feas ≈ b)
│ println("And inside bound constraints: ", all(-10.0 .<= result.x_best_feas .<= 10.0))
│ ```
│   c.value = UndefVarError: NomadProblem not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/FuXcO/src/Expanders.jl:563
[ Info: CrossReferences: building cross-references.
[ Info: CheckDocument: running document checks.
[ Info: Populate: populating indices.
[ Info: RenderDocument: rendering document.
[ Info: HTMLWriter: rendering HTML pages.
```

File: src/converters.jl [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+function convert_to_z(c::AbstractConverter, x::Vector{Float64})::Vector{Float64} end
+
+function get_nz_dimension(c::AbstractConverter)::Int end
+
+struct SVDConverter <: AbstractConverter
+
+    U::Array{Float64, 2} # U factor
+    S::Vector{Float64} # Singular values
+    Vt::Array{Float64, 2} # V factor
```
Comment (amontoison) [inline src/converters.jl:123]: `Matrix{Float64}` ?
`U` and `Vt` are always dense matrices ?

File: src/core.jl [hunk 6]
```diff
@@ -287,7 +358,8 @@ The problem to solve.
 - `x0::Vector{Float64}`
 The starting point. Must satisfy lb <= x0 <= ub
 where lb and ub are respectively the lower and upper bounds
-of the NomadProblem p.
+of the NomadProblem p. When A and b are defined, it must satisfy
+Ax0=b.
```
Comment (amontoison) [inline src/core.jl:362]: `A * x0 = b`

**Review by amontoison — CHANGES_REQUESTED** [review]
Great work !!! Thanks Ludovic.
I have few minor comments.

File: src/converters.jl (hunk 5 above)
Comment (salomonl) [reply src/converters.jl:123]: Obviously, some performance could be gained if we analyse more the structure of the matrix. But in blackbox optimization, most of the time, the number of variables is small (max 60), and so the computation cost to compute the SVD or QR factorization is negligeable relatively to the cost of the blackbox.

Comment (amontoison) [reply src/converters.jl:123]: Perfect, thanks for the clarification Ludovic !

File: docs/src/tutorial.md (hunk 4 above)
Comment (salomonl) [reply docs/src/tutorial.md:135]: I do not understand. Normally, it should not take into consideration `src/nomadParameters.md` as it is not included in `make.jl`

Comment (amontoison) [reply docs/src/tutorial.md:135]: It's `docs/src/nomadParameters.md` that is not included in `make.jl` and not `src/nomadParameters.md`.

File: Project.toml (hunk 1 above)
Comment (salomonl) [reply Project.toml]: I updated to julia^1.5 equally.

File: Project.toml [hunk 7]
```diff
@@ -11,4 +11,5 @@ Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 Tulip = "6dd1b50a-3aae-11e9-10b5-ef983d2400fa"
 
 [compat]
-julia = "^1.3"
+julia = "^1.5"
```
Comment (amontoison) [inline Project.toml:14]: Laisse `^1.3` vu que les artifacts marchent depuis cette version. On a aucune raison de limiter à la version 1.5 non?

**Comment by amontoison** [comment]
I will do a new release if it's fine for you @salomonl.

I also give you the commands to rebase your fork's branchs with the bbopt/NOMAD.jl repository:
```
git remote add upstream https://github.com/bbopt/NOMAD.jl.git
git fetch upstream
git checkout my_branch
git rebase upstream/master
git push -f
```

**Comment by salomonl** [comment]
Ok for that. It is a minor release, as it only adds some new features.

**Comment by amontoison** [comment]
Ok I will  release the version `2.0.1`.
I just need to fix the deployment of the documentation before.

**Pull request merged**